lib/puppet/resource_api.rb

```diff
@@ -178,9 +178,14 @@ def to_resource
             isnamevar
           end
 
+          type = Puppet::ResourceApi.parse_puppet_type(name, options[:type])
+
           # read-only values do not need type checking, but can have default values
-          if options[:behaviour] != :read_only
-            if options.key? :default
+          if options[:behaviour] != :read_only && options.key?(:default)
+            if type.is_a? Puppet::Pops::Types::PBooleanType
```

Comment:
What about compound types like `Optional[Boolean]`, `Array[Boolean]` and `Variant[String, Boolean]`?

Reply:
Would it be enough to apply the workaround if and only if the default value is actually `false`?